File: projects/Maths for DSA/Basic Maths for DSA - Striver's A2Z DSA Course/Print all Divisors of a Number/optimizedApproach.js
#!/bin/env node
/**
 * Print all divisors from a given Number:
 *
 *  Problem Statement:
 *  Given an integer n, print all divisors of n in increasing order.
 *  A divisor of n is any integer d such that n % d == 0.
 */

/**
 *
 * @param {*} number - Target to found all Divisors of
 * @returns [int] - Divisors
 */
function printDivisors(number) {
  let res = [];
  let iterator = 1;

  while (iterator <= Math.sqrt(number)) {
    if (number % iterator === 0) {
      res.push(iterator);
      if (number / iterator !== iterator) {
        res.push(number / iterator);
      }
    }
    iterator++;
  }

  return res.sort((a, b) => a - b);
}

// Test Cases.
console.log("Test Case 36");
console.log(printDivisors(36));

console.log();
console.log("Test Case 80");
console.log(printDivisors(80));

File: DSA/maths for DSA/print all divisors of a number/optimized_approach.py
#!/bin/python3
"""
    Print all divisors from a given Number:

    Problem Statement:
    Given an integer n, print all divisors of n in increasing order.
    A divisor of n is any integer d such that n % d == 0.
"""

from math import sqrt

def print_divisors(number):
    """Naive Solution Approach."""
    res = []
    iterator = 1

    while iterator <= sqrt(number):
        if number % iterator == 0:
            res.append(iterator)
            if number // iterator != iterator:
                res.append(number // iterator)
        iterator += 1

    return sorted(res)

if __name__ == "__main__":
    print("Test Case 36")
    print(print_divisors(36))

    print()
    print("Test Case 80")
    print(print_divisors(80))
